image/cas/layout/tar.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package layout
+
+import (
+	"archive/tar"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/opencontainers/image-spec/image/cas"
+	"github.com/opencontainers/image-spec/image/layout"
+)
+
+// TarEngine is a cas.Engine backed by a tar file.
+type TarEngine struct {
+	reader ReadSeekCloser
+}
+
+// NewTarEngine returns a new TarEngine.
+func NewTarEngine(file ReadSeekCloser) (eng cas.Engine, err error) {
+	engine := &TarEngine{
+		reader: file,
+	}
+
+	err = layout.CheckTarVersion(engine.reader)
+	if err != nil {
+		return nil, err
+	}
+
+	return engine, nil
+}
+
+// Put adds a new blob to the store.
+func (engine *TarEngine) Put(reader io.Reader) (digest string, err error) {
+	// FIXME
+	return "", errors.New("TarEngine.Put is not supported yet")
+}
+
+// Get returns a reader for retrieving a blob from the store.
+func (engine *TarEngine) Get(digest string) (reader io.ReadCloser, err error) {
+	fields := strings.SplitN(digest, ":", 2)
+	if len(fields) != 2 {
+		return nil, fmt.Errorf("invalid digest: %q, %v", digest, fields)
+	}
+	algorithm := fields[0]
+	hash := fields[1]
+
+	targetName := fmt.Sprintf("./blobs/%s-%s", algorithm, hash)
+
+	_, err = engine.reader.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return nil, err
+	}
+
+	tarReader := tar.NewReader(engine.reader)
+	for {
+		header, err := tarReader.Next()
```

Comment:
Rather than doing an `O(N)` search through the tar every time, it is sufficient to make a single pass over the tar, noting the offset and length of the target file. When searching for the file in Get, this becomes and index lookup and seek, rather than having to scan the whole file.

This obviously becomes more tricky with compression, so temporary files may be a better approach.

@vbatts may have more detailed information on doing this correctly.